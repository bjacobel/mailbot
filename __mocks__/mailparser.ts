import { Transform } from "stream";
import { inherits } from "util";

export const MailParser = jest.fn(function () {
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore 2683
  Transform.call(this, { objectMode: true });
});
inherits(MailParser, Transform);

interface TextData {
  [name: string]: string;
  type: "text";
}

MailParser.prototype._transform = jest.fn(
  (
    data: string,
    encoding: string,
    callback: (err: Error | undefined, data: TextData) => void,
  ) => {
    callback(undefined, { html: data.toString(), type: "text" });
  },
);
