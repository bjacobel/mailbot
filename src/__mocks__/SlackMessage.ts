// tslint:disable no-empty
const SlackMessage = jest.fn();

SlackMessage.prototype._transform = jest.fn(() => {});

export default SlackMessage;
