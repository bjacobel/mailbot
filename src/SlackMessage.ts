import { ChatPostMessageArguments /* WebClient */ } from "@slack/web-api";
import { EmailAddress, HeaderValue } from "mailparser";

export type Headers = Map<string, string | HeaderValue>;

export default class SlackMessage {
  private body: string;
  private headers: Headers;
  // private slack: WebClient;

  constructor(emailBody: string, emailHeaders: Headers) {
    this.body = emailBody;
    this.headers = emailHeaders;
    // this.slack = new WebClient();
  }

  public async composeMessage(): Promise<ChatPostMessageArguments> {
    return {
      mrkdwn: true,
      username: this.fullname(),
      channel: "",
      text: this.body, // todo: transform html to markdown
    };
  }

  public async send(): Promise<void> {
    return Promise.resolve();
    // const args = await this.composeMessage();
    // log.info(JSON.stringify(args));
    // this.slack.chat.postMessage({
    //   ...args,
    // });
  }

  public fullname(): string {
    if (!this.headers.has("from")) {
      return "Unknown sender";
    }

    const fromHeader = this.headers.get("from");
    if (typeof fromHeader === "string") {
      return fromHeader.split("<")[0].trim();
    }

    if (typeof fromHeader === "object" && "value" in fromHeader) {
      return (fromHeader.value[0] as EmailAddress).name;
    }

    return "Unknown sender";
  }

  public async avatar(): Promise<string> {
    return Promise.resolve("");
  }
}
