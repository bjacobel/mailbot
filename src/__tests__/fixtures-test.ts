jest.dontMock("mailparser");
jest.dontMock("axios");

import * as glob from "glob";
import { HeaderValue, simpleParser } from "mailparser";
import * as path from "path";

import SlackMessage, { Headers } from "../SlackMessage";

describe("fixtures", () => {
  const fixtureDir = path.join(__dirname, "../../fixtures/email");
  glob.sync(path.join(fixtureDir, "**/*.txt")).forEach((fixture) => {
    const relpath = path.relative(__dirname, fixture);

    it.call(
      global,
      path.relative(fixtureDir, fixture),
      async () => {
        const content = await import(relpath);

        const parsedEmail = await simpleParser(content);
        const { html } = parsedEmail;
        const headers: Headers = new Map(
          [...parsedEmail.headers.entries()].map(
            ([key, value]: [string, HeaderValue]) =>
              [key, value] as [string, string],
          ),
        );

        const slack = new SlackMessage(html as string, headers);
        expect(await slack.composeMessage()).toMatchSnapshot();
      },
      10000,
    );
  });
});
