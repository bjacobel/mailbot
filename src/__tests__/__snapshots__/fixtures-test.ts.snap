// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures bulletin.txt 1`] = `
Object {
  "channel": "",
  "mrkdwn": true,
  "text": false,
  "username": "UpperValley DSA",
}
`;
